File: bindings/android/hello-scion/src/main/kotlin/com/anapaya/scion/http3/hello/LocalNetwork.kt
// Copyright 2026 Anapaya Systems

package com.anapaya.scion.http3.hello

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.URL

/**
 * Where the local test network is, and how to be trusted by it.
 *
 * A real app knows all of this before it is built: a fixed endhost API URL, a token from where it
 * keeps credentials, a server it addresses by name, and an authority it ships or inherits from the
 * device. `scion-h3-test-server` decides all four when it starts, because it takes ephemeral ports,
 * mints an auth token per run and generates a throwaway certificate authority, so this sample asks
 * it.
 */
data class LocalNetwork(
    val endhostApiUrl: String,
    val authToken: String,
    val baseUrl: String,
    /** The server's SCION address, without a port: the port comes from [baseUrl]. */
    val target: String,
    val caPem: String,
) {
    companion object {
        /**
         * The control API of a test server started with `--control-port 7443`. 10.0.2.2 is the
         * emulator's translation of the host's loopback, which is where the server binds.
         */
        private const val CONTROL_URL = "http://10.0.2.2:7443/info"

        /** Reads the description the test server serves. Throws if it is not running. */
        suspend fun discover(): LocalNetwork =
            withContext(Dispatchers.IO) {
                val info = JSONObject(URL(CONTROL_URL).readText())
                LocalNetwork(
                    endhostApiUrl = info.getString("endhost_api_url"),
                    authToken = info.getString("auth_token"),
                    baseUrl = info.getString("base_url"),
                    target = info.getString("target"),
                    caPem = info.getString("ca_pem"),
                )
            }
    }
}

File: bindings/apple/hello-scion/HelloScion/LocalNetwork.swift
// Copyright 2026 Anapaya Systems

import Foundation

/// Where the local test network is.
///
/// A real app knows all of this before it is built. `scion-h3-test-server` decides all four when it
/// starts, because it takes ephemeral ports, mints an auth token per run and generates a throwaway
/// certificate authority.
struct LocalNetwork: Decodable, Sendable {
    let endhostApiUrl: String
    let authToken: String
    let baseUrl: String
    let target: String
    let caPem: String

    enum CodingKeys: String, CodingKey {
        case endhostApiUrl = "endhost_api_url"
        case authToken = "auth_token"
        case baseUrl = "base_url"
        case target
        case caPem = "ca_pem"
    }

    /// The control API of a test server started with `--control-port 7443`. The simulator shares
    /// the Mac's network, so the loopback address the server binds is the simulator's too.
    private static let controlUrl = "http://127.0.0.1:7443/info"

    /// Reads the description the test server serves. Throws if it is not running.
    static func discover() async throws -> LocalNetwork {
        let (data, _) = try await URLSession.shared.data(from: URL(string: controlUrl)!)
        return try JSONDecoder().decode(LocalNetwork.self, from: data)
    }
}
